File: frontend/src/api/documents.ts
import { API_BASE_URL, apiRequest } from "./client";

export type DocumentKind = "supplier" | "esg" | "transaction";
export type IngestionStatus = "queued" | "processing" | "completed" | "failed";
export type IngestionStepStatus = "pending" | "processing" | "completed" | "failed";

export interface UploadUrlRequest {
  fileName: string;
  documentKind: DocumentKind;
  contentType: string;
}

export interface UploadUrlResponse {
  fileName: string;
  documentKind: DocumentKind;
  blobKey: string;
  uploadUrl: string;
  uploadMethod: "PUT" | "POST";
  headers?: Record<string, string>;
}

export interface IngestionCreateFile {
  fileName: string;
  documentKind: DocumentKind;
  blobKey: string;
}

export interface IngestionCreateRequest {
  files: IngestionCreateFile[];
}

export interface IngestionStep {
  name: "upload" | "extract" | "parse" | "persist" | "history";
  status: IngestionStepStatus;
  message?: string;
}

export interface IngestionSummary {
  supplierRecordsAdded: number;
  esgRecordsAdded: number;
  transactionRecordsAdded: number;
}

export interface IngestionJob {
  jobId: string;
  status: IngestionStatus;
  createdAt: string;
  updatedAt: string;
  steps: IngestionStep[];
  summary?: IngestionSummary;
}

export interface DocumentHistoryItem {
  documentName: string;
  documentType: string;
  status: string;
  recordsAdded: string;
  timestamp: string;
}

export async function getUploadUrl(payload: UploadUrlRequest): Promise<UploadUrlResponse> {
  return apiRequest<UploadUrlResponse>("/documents/upload-url", {
    method: "POST",
    json: payload,
  });
}

export async function uploadDocumentToStorage(
  file: File,
  uploadConfig: UploadUrlResponse,
): Promise<void> {
  const uploadUrl = new URL(uploadConfig.uploadUrl, `${API_BASE_URL}/`).toString();

  const response = await fetch(uploadUrl, {
    method: uploadConfig.uploadMethod,
    headers: {
      "Content-Type": file.type || "application/pdf",
      ...(uploadConfig.headers ?? {}),
    },
    body: file,
  });

  if (!response.ok) {
    throw new Error(`Upload failed for ${file.name}`);
  }
}

export async function createIngestionJob(
  payload: IngestionCreateRequest,
): Promise<IngestionJob> {
  return apiRequest<IngestionJob>("/documents/ingestions", {
    method: "POST",
    json: payload,
  });
}

export async function getIngestionJob(jobId: string): Promise<IngestionJob> {
  return apiRequest<IngestionJob>(`/documents/ingestions/${jobId}`);
}

export async function getDocumentHistory(): Promise<DocumentHistoryItem[]> {
  return apiRequest<DocumentHistoryItem[]>("/documents/history");
}

File: backend/app/schemas/documents.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


DocumentKind = Literal["supplier", "esg", "transaction"]
IngestionStatus = Literal["queued", "processing", "completed", "failed"]
StepStatus = Literal["pending", "processing", "completed", "failed"]
UploadMethod = Literal["PUT", "POST"]


class UploadUrlRequest(BaseModel):
    fileName: str = Field(min_length=1)
    documentKind: DocumentKind
    contentType: str = Field(min_length=1)


class UploadUrlResponse(BaseModel):
    fileName: str
    documentKind: DocumentKind
    blobKey: str
    uploadUrl: str
    uploadMethod: UploadMethod
    headers: dict[str, str] | None = None

    model_config = ConfigDict(from_attributes=True)


class IngestionCreateFile(BaseModel):
    fileName: str = Field(min_length=1)
    documentKind: DocumentKind
    blobKey: str = Field(min_length=1)


class IngestionCreateRequest(BaseModel):
    files: list[IngestionCreateFile] = Field(min_length=1)


class IngestionStep(BaseModel):
    name: Literal["upload", "extract", "parse", "persist", "history"]
    status: StepStatus
    message: str | None = None


class IngestionSummary(BaseModel):
    supplierRecordsAdded: int
    esgRecordsAdded: int
    transactionRecordsAdded: int


class IngestionJobResponse(BaseModel):
    jobId: str
    status: IngestionStatus
    createdAt: datetime
    updatedAt: datetime
    steps: list[IngestionStep]
    summary: IngestionSummary | None = None


class DocumentHistoryItem(BaseModel):
    documentName: str
    documentType: str
    status: str
    recordsAdded: str
    timestamp: str
